### DISCUSSION TITLE
Prescribed motion at tower tip for coupling with external solvers

### DISCUSSION THREAD

**Discussion by Tudore** [OpenFAST/openfast#901] [Q&A]
Dear Dr. Jonkman,

Thanks for bringing FAST and OpenFAST to us, which are very powerful tools for the community!

I am a researcher from Singapore. My colleague and I just started to look into the source code of OpenFAST. We wish to achieve a two-way coupling that we provide prescribed motions to the turbine at the tip of the tower with the motions calculated/modeled in external solvers or tests. In other words, the motions below the tip of the tower will be prescribed rather than calculated by OpenFAST. Meantime, the reaction forces from the turbine on the tower tip will be transferred back to the external solvers/ tests. 

Based on our understanding after reading the codes and your reply in the discussions on "external FE  coupling", we may need to modify the ElastoDyn module to achieve the function. We noted your constructive suggestions to @ZahidUllah33400: since the displacement-related variables are known before the matrix solving, we need to modify the matrix of motion equation as in  the following figure (cited from your answer):       

![image](https://user-images.githubusercontent.com/35771437/139533405-391c77fd-474c-4b22-a589-0b2c906dc79f.png)

We also noted that the displacement of the platform is solved by external code in the abovementioned discussion. So, only the 6-DOF motions expressed in the matrix needs modifications. For our case, taking the assumption that the tower and the platform will not be calculated in OpenFAST, the number of unknown DOF in the equation should be reduced from 24 to 14. The reduced DOFs are 6 for the floating platform and 4 for the tower. These DOFs are either known to be zero or prescribed.  

We propose the following path for our code development:

Firstly, we force the motion of the platform and tower to be zero during the simulation, namely, they are stationary. However, the tower tip node will still have motion. The motion will be imported from external sources. To achieve this, we hope to modify the variables related to ED_MiscVarType%RtHS and ED_MiscVarType%CoordSys. This is to make sure the variables that affect the mass matrix, the force vector of the motion equation and the output of the ED are kept our desired values. However, we noted that the modal superposition method is used in the solver, we are not quite sure how to make sure tower tip motion can be transfered to the code. This approach possibly will affect the generalized mass matrix and generalized force vector when solving the tower dynamics. 

If the approach we proposed is not feasible, may I seek your suggestion on how to achieve the function we are looking forward to? We will be grateful for your kind help and suggestions.

Regards,
Chi

**Comment by jjonkman** [comment]
Dear @Tudore,

If I understand correctly, you want OpenFAST to model the rotor-nacelle assembly (RNA) while the support structure (including tower) are modeled in an external solver.  You can actually follow the approach I explained to @ZahidUllah33400 in https://github.com/OpenFAST/openfast/discussions/801, with one modification.  Essentially, you'll want to locate the platform in ElastoDyn at the tower top and eliminate the tower model and degrees of freedom in ElastoDyn.  This can be achieved by setting inputs in the ElastoDyn input file such that terms associated with the tower are zero or negligible.  That is, you should set `PtfmRefzt` in ElastoDyn to the height of the tower top, disable the tower degrees of freedom, set the geometry so that tower length is zero, and set the tower mass to zero or negligible values.  Effectively, the 6 DOFs of the platform are then at the tower top and there is no tower modeled in ElastoDyn.  And you can then follow the approach I describe in #801 where the 6-DOF motion of the platform is prescribed instead of calculated.

Best regards,
